src/ChannelServer/World/SkillHandler.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Melia.Channel.Network;
+using Melia.Shared.Data.Database;
+using Melia.Shared.World;
+
+namespace Melia.Channel.World
+{
+	interface ITargetedSkillHandler
+	{
+		void Handle(Skill skill, Character caster, IEntity target);
+	}
+
+	interface IGroundSkillHandler
+	{
+		void Handle(Skill skill, Character caster, Position position);
+	}
+
+	interface ITargetGroundSkillHandler
+	{
+		void Handle(Skill skill, Character caster, Position position, List<IEntity> targets);
+	}
+
+	class TargetedSkillHandler : ITargetedSkillHandler
+	{
+		public void Handle(Skill skill, Character caster, IEntity target)
+		{
+			if (skill.Data.OverHeat != 0)
+				skill.OverHeatCount++;
+			Send.ZC_OVERHEAT_CHANGED(caster, skill);
+			if (skill.Data.OverHeat == 0 || skill.Data.OverHeat == skill.OverHeatCount)
+			{
+				Send.ZC_COOLDOWN_CHANGED(caster, skill);
+				skill.OverHeatCount = 0;
+			}
+
+			var damage = caster.GetRandomPAtk() + 100;
+			Send.ZC_SKILL_FORCE_TARGET(caster, target, skill, damage);
+			// Unsafe cast for now we're only dealing with monsters
+			if (target != null)
+			{
+				var monster = (Monster)target;
+				if (monster.TakeDamage(damage, caster, 0))
+					Send.ZC_SKILL_CAST_CANCEL(caster, target);
+			}
+		}
+	}
+
+	class TargetedGroundSkillHandler : ITargetGroundSkillHandler
+	{
+		public void Handle(Skill skill, Character caster, Position position, List<IEntity> targets)
+		{
+			if (skill.Data.OverHeat != 0)
+				skill.OverHeatCount++;
+			Send.ZC_OVERHEAT_CHANGED(caster, skill);
+			if (skill.Data.OverHeat == 0 || skill.Data.OverHeat == skill.OverHeatCount)
```

Comment:
Instead of manually checking variables, it would be clearer to add properties and methods to do these things. For example, this would be much more descriptive and easier to use:
```cs
if (skill.IsOverheated)
```
Otherwise, you have to guess a little what the code is supposed to do, which I did here, I'm not actually sure `IsOverheated` would be correct^^